### PR TITLE
CORE-908 Version of Node that uses InMemory{Store,TrieStore}

### PR DESCRIPTION
**Pull request by ys-pyrofex**
## Overview
InMemory RSpace will be useful for performance testing and possible launch of test node on systems where LMDBStore isn't yet supported (WSL on Windows for example)

### Which JIRA issue does this PR relate to? 
https://rchain.atlassian.net/browse/CORE-908

### CODE REVIEW

File: rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala [hunk 1]
```diff
@@ -344,9 +344,17 @@ object ReplayRSpace {
     implicit val codecA: Codec[A] = sa.toCodec
     implicit val codecK: Codec[K] = sk.toCodec
 
-    val mainStore = LMDBStore.create[C, P, A, K](context, branch)
-
-    val replaySpace = new ReplayRSpace[C, P, A, R, K](mainStore, branch)
+    val (mainStore, replaySpace) = context match {
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]: The code duplication below is unnecessary

File: rspace/src/main/scala/coop/rchain/rspace/InMemoryStore.scala [hunk 2]
```diff
@@ -261,7 +261,7 @@ object InMemoryStore {
     implicit val codecC: Codec[C] = sc.toCodec
     implicit val codecP: Codec[P] = sp.toCodec
     implicit val codecA: Codec[A] = sa.toCodec
-    initialize(trieStore, branch)
+
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/InMemoryStore.scala]: Why is this not needed anymore?

File: docs/windows/rnode.toml [hunk 3]
```diff
@@ -10,6 +10,7 @@ bootstrap = "rnode://c61769b39d368cbcbc9499634e030386c79d5b02@52.119.8.108:40400
 standalone = true
 data-dir = "C:/RChain/data"
 map-size = 200000000
+in-memory-store = true
```
Comment (lukasz-golebiewski) [inline docs/windows/rnode.toml]: Why is true default for Windows?

**Comment by dzajkowski** [comment]
Looks good enough to me, please also ask @rabbitonweb for node changes.

File: rspace/src/main/scala/coop/rchain/rspace/InMemoryStore.scala (hunk 2 above)
Comment (ys-pyrofex) [reply rspace/src/main/scala/coop/rchain/rspace/InMemoryStore.scala]: With latest update InMemoryStore implementation is closer to LMDBStore, which doesn't need this call as well.

File: rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala (hunk 1 above)
Comment (ys-pyrofex) [reply rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]: You're right, fixed.

File: docs/windows/rnode.toml (hunk 3 above)
Comment (ys-pyrofex) [reply docs/windows/rnode.toml]: Mistakenly left from my local benchmarks. Set to false.

File: docs/windows/rnode.toml [hunk 4]
```diff
@@ -10,6 +10,7 @@ bootstrap = "rnode://c61769b39d368cbcbc9499634e030386c79d5b02@52.119.8.108:40400
 standalone = true
 data-dir = "C:/RChain/data"
 map-size = 200000000
+in-memory-store = false
```
Comment (rabbitonweb) [inline docs/windows/rnode.toml]: this configuration file is for windows, should it not be set to true?

File: rspace/src/main/scala/coop/rchain/rspace/Context.scala [hunk 5]
```diff
@@ -3,23 +3,35 @@ package coop.rchain.rspace
...
   def close(): Unit = {
     trieStore.close()
     env.close()
   }
 }
 
+class InMemoryContext[C, P, A, K] private[rspace] (
+    val trieStore: ITrieStore[InMemTransaction[history.State[Blake2b256Hash, GNAT[C, P, A, K]]],
+                              Blake2b256Hash,
+                              GNAT[C, P, A, K]]
```
Comment (rabbitonweb) [inline rspace/src/main/scala/coop/rchain/rspace/Context.scala]: "GNAT C.P.A.K" sounds like a NY rap group :)

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 6]
```diff
@@ -189,10 +189,16 @@ object RSpace {
       sc: Serialize[C],
       sp: Serialize[P],
       sa: Serialize[A],
-      sk: Serialize[K]): RSpace[C, P, A, R, K] = {
-    val mainStore = LMDBStore.create[C, P, A, K](context, branch)
-    create(mainStore, branch)
-  }
+      sk: Serialize[K]): RSpace[C, P, A, R, K] =
+    context match {
```
Comment (rabbitonweb) [inline rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: Just so I understand this correctly (sorry if asking dumb question. We are given here a `context` as parameter and we create one again?

File: rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala [hunk 7]
```diff
@@ -344,7 +344,13 @@ object ReplayRSpace {
     implicit val codecA: Codec[A] = sa.toCodec
     implicit val codecK: Codec[K] = sk.toCodec
 
-    val mainStore = LMDBStore.create[C, P, A, K](context, branch)
+    val mainStore = context match {
```
Comment (rabbitonweb) [inline rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]: same question as above

File: docs/windows/rnode.toml (hunk 4 above)
Comment (ys-pyrofex) [reply docs/windows/rnode.toml]: On Windows (not WSL) LMDB works fine. The only problem we have now - absent native DLLs in our port of kalium. In memory store should be default for WSL, but we don't have a sample config for WSL yet.

File: rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala (hunk 7 above)
Comment (ys-pyrofex) [reply rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]: Replaced "match" with context.createStore method.

**Pull request merged**